File: python/498.py
"""Project Euler Problem 498: Remainder of polynomial division.

Find the coefficient of x^D in the remainder when x^N is divided by (x-1)^K.
"""

from __future__ import annotations


def ncr_small(n: int, k: int, mod: int) -> int:
    """nCk modulo prime mod with small k."""
    if k < 0 or k > n:
        return 0
    k = min(k, n - k)
    if k == 0:
        return 1
    num = 1
    den = 1
    for i in range(1, k + 1):
        num = (num * (n - k + i)) % mod
        den = (den * i) % mod
    return num * pow(den, mod - 2, mod) % mod


def ncr_lucas(n: int, k: int, mod: int) -> int:
    """Lucas theorem for prime mod."""
    if k < 0 or k > n:
        return 0
    result = 1
    while n > 0 or k > 0:
        ni = n % mod
        ki = k % mod
        if ki > ni:
            return 0
        result = (result * ncr_small(ni, ki, mod)) % mod
        n //= mod
        k //= mod
    return result


def solve() -> int:
    """Solve Problem 498."""
    N = 10**13
    K = 10**12
    D = 10**4
    M = 999_999_937
    n1 = N - D - 1
    k1 = K - 1 - D
    coeff = ncr_lucas(N, D, M)
    coeff = (coeff * ncr_lucas(n1, k1, M)) % M
    return coeff


def main() -> int:
    """Main entry point."""
    result = solve()
    print(result)
    return result


if __name__ == "__main__":
    main()

File: c/498.c
/* Project Euler 498 - Remainder of polynomial division
 * Translated from python/498.py
 *
 * Coefficient of x^D in remainder when x^N is divided by (x-1)^K.
 * Uses Lucas theorem for nCk mod prime.
 */
#include <stdio.h>
#include <stdint.h>

typedef long long ll;
typedef unsigned long long ull;
typedef __int128 lll;

#define MOD 999999937LL

ll mulmod(ll a, ll b, ll m) {
    return (lll)a * b % m;
}

ll powmod(ll base, ll exp, ll m) {
    ll result = 1;
    base %= m;
    if (base < 0) base += m;
    while (exp > 0) {
        if (exp & 1) result = mulmod(result, base, m);
        base = mulmod(base, base, m);
        exp >>= 1;
    }
    return result;
}

/* nCk mod prime, small k (k < mod) */
ll ncr_small(ll n, ll k, ll mod) {
    if (k < 0 || k > n) return 0;
    if (k > n - k) k = n - k;
    if (k == 0) return 1;
    ll num = 1, den = 1;
    for (ll i = 1; i <= k; i++) {
        num = mulmod(num, (n - k + i) % mod, mod);
        den = mulmod(den, i % mod, mod);
    }
    return mulmod(num, powmod(den, mod - 2, mod), mod);
}

/* Lucas theorem for nCk mod prime */
ll ncr_lucas(ll n, ll k, ll mod) {
    if (k < 0 || k > n) return 0;
    ll result = 1;
    while (n > 0 || k > 0) {
        ll ni = n % mod;
        ll ki = k % mod;
        if (ki > ni) return 0;
        result = mulmod(result, ncr_small(ni, ki, mod), mod);
        n /= mod;
        k /= mod;
    }
    return result;
}

int main() {
    ll N = 10000000000000LL;   /* 10^13 */
    ll K = 1000000000000LL;    /* 10^12 */
    ll D = 10000LL;            /* 10^4 */

    ll n1 = N - D - 1;
    ll k1 = K - 1 - D;

    ll coeff = ncr_lucas(N, D, MOD);
    coeff = mulmod(coeff, ncr_lucas(n1, k1, MOD), MOD);

    printf("%lld\n", coeff);
    return 0;
}
